### PR TITLE
Feature/improvements

### PR DESCRIPTION
**Pull request by ismi29ch**
- Added support for typescript
- Handle edge cases (error handling)
- Auto Pick name from Url
- User defined delay (Allow to pass custom delay)

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for creating a TypeScript definition. 🙌 Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)?

Some examples:
- https://github.com/sindresorhus/filled-array/commit/aae7539cb32f163cb063499664b012d0b04b3104
- https://github.com/sindresorhus/write-json-file/blob/main/index.d.ts
- https://github.com/sindresorhus/delay/blob/main/index.d.ts

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	 */
+	export default async function multiDownload(
+		urls: string[],
+		options?: {
+			rename?: (options: {
+				url: string;
+				index: number;
+				urls: string[];
+			}) => string;
+			delayTime?: number;
```
Comment (sindresorhus) [inline index.d.ts:17]: The name does not make it clear that it's a delay between each download.

You also need to add it to the readme.

**Comment by ismi29ch** [comment]
Hey @sindresorhus,
I've made the requested changes, please look into them & let me know if still there's a room for improvement. Thanks

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	 * The function to rename the downloaded files.
+	 * @default undefined (use original filenames)
+	 */
+	rename?: RenameFunction;
+
+	/**
+	 * The delay time between each file download in milliseconds.
+	 * @default 1000
+	 */
+	delayTime?: number;
```
Comment (sindresorhus) [inline index.d.ts:27]: As commented earlier, this needs a better name.

File: readme.md [hunk 3]
```diff
@@ -74,6 +74,12 @@ Type: `Function`
 
 A function that accepts an object containing `url`, `index`, and `urls` properties and is expected to return the new filename.
 
+##### delayTime
+
+Type: `number`
+
+Time in miliseconds to delay between each download
```
Comment (sindresorhus) [inline readme.md:81]: This is not in sync with index.d.ts

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Please rebase from the `main` branch and fix any linting issues.

**Comment by ismi29ch** [comment]
Hey @sindresorhus,
I've made the requested changes again
please look into them & let me know if still there's a room for improvement. Thanks

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  Download multiple files from an array of URLs.
+  @param urls - The URLs to download.
+  @param options - The options for downloading.
+  @returns A promise that resolves when all files have been downloaded.
+  @example
+  ```
+  import { multiDownload } from 'multi-download';
+
+  async function downloadFiles() {
+    const urls = [
```
Comment (sindresorhus) [inline index.d.ts:35]: Use tab-indentation for the code example.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+/**
+  Download multiple files from an array of URLs.
+  @param urls - The URLs to download.
+  @param options - The options for downloading.
+  @returns A promise that resolves when all files have been downloaded.
+  @example
+  ```
+  import { multiDownload } from 'multi-download';
```
Comment (sindresorhus) [inline index.d.ts:32]: Follow the existing code style.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  This function is used to determine the filename for each downloaded file.
+  @param options - The options for renaming.
+  @returns The new filename for the downloaded file.
+*/
+export type RenameFunction = (options: {url: string; index: number; urls: string[]}) => string;
+
+/*
+  Options for the multiDownload function.
+ */
+export type MultiDownloadOptions = {
```
Comment (sindresorhus) [inline index.d.ts:12]:
```suggestion
export type Options = {
```

File: index.js [hunk 7]
```diff
@@ -15,15 +16,32 @@ const download = async (url, name) => {
...
 
-	for (const [index, url] of urls.entries()) {
-		const name = typeof rename === 'function' ? rename({url, index, urls}) : '';
+	if (urls.some(url => typeof url !== 'string')) {
+		throw new Error('`urls` must be an array of strings');
+	}
 
-		await delay(index * 1000); // eslint-disable-line no-await-in-loop
+	if (typeof downloadInterval !== 'number') {
+		throw new TypeError('`downloadInterval` must be a number');
```
Comment (sindresorhus) [inline index.js:35]:
```suggestion
		throw new TypeError('The `downloadInterval` option must be a number');
```

File: index.js [hunk 8]
```diff
@@ -15,15 +16,32 @@ const download = async (url, name) => {
...
+	}
+
+	if (typeof rename !== 'function') {
+		throw new TypeError('`rename` must be a function');
```
Comment (sindresorhus) [inline index.js:39]:
```suggestion
		throw new TypeError('The `rename` option must be a function');
```

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "multi-download",
-	"version": "4.0.0",
+	"version": "4.1.0",
```
Comment (sindresorhus) [inline package.json:3]: I already asked you not to bumo the version.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 4 above)
Comment (ismi29ch) [reply index.d.ts:35]: seems already tab indend

**Comment by ismi29ch** [comment]
@sindresorhus 
Its been a while can you look into the PR again?

File: readme.md [hunk 10]
```diff
@@ -74,6 +74,12 @@ Type: `Function`
 
 A function that accepts an object containing `url`, `index`, and `urls` properties and is expected to return the new filename.
 
+##### downloadInterval
+
+Type: `number`
+
+The delay time between each file download in milliseconds.
```
Comment (sindresorhus) [inline readme.md:81]: This is the wrong location. Please look at the readme and you'll see that the below code example is for the `rename` option.

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  		console.log("Files downloaded successfully");
+  	} catch (error) {
+  		console.error("Error downloading files:", error);
+  	}
+
+  }
+
+  await downloadFiles();
+  ```
+ */
```
Comment (sindresorhus) [inline index.d.ts:66]: This file is messy code style wise. You are mixing spaces with tabs and the formatting is completely off.

Use the formatting shown in https://github.com/sindresorhus/typescript-definition-style-guide#documentation

Use single-quotes.

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
+/**
+  @param options   -  The options for renaming.
+
+  @returns         -  The new filename for the downloaded file.
+
+  @description     -  This function is used to determine the filename for each downloaded file.
+*/
+export type RenameFunction = (options: {
```
Comment (sindresorhus) [inline index.d.ts:8]: `options` is the wrong term here. They are not options.

File: index.d.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	 */
+	downloadInterval?: number;
+};
+
+/**
+  @param urls     - The URLs to download.
+
+  @param options  - The options for downloading.
+
+  @returns        - A promise that resolves when all files have been downloaded.
```
Comment (sindresorhus) [inline index.d.ts:36]: Use the correct TSDoc syntax. Applies to the everything else too.

```suggestion
  @returns A promise that resolves when all files have been downloaded.
```

File: index.d.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+}) => string;
+
+/*
+  Options for the multiDownload function.
+ */
+export type Options = {
+	/**
+	  @default undefined  - (use original filenames)
+
+	  @description        - A function that returns the new filename for each downloaded file.
```
Comment (sindresorhus) [inline index.d.ts:21]: Missing example like in the readme.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Pull request closed without merging**